### PR TITLE
[Enhancement][Vectorized] Improve hash table build efficiency

### PR DESCRIPTION
**Pull request by jacktengg**
# Proposed changes
1. MAP_POPULATE is missing for mmap in Allocator, because macro OS_LINUX is not defined in allocator.h;
2. MAP_POPULATE has no effect for mremap as for mmap, zero-fill enlarged memory range explicitly to pre-fault the pages

After these improvement, queries for ssb 100G have about 14% total performance enhancement.
Test Config:
4 servers(SSD 1T, 16 core CPU，64G RAM），1fe、3be
Config parameters:
```
enable_vectorized_engine=1;
parallel_fragment_exec_instance_num=8;
exec_mem_limit=50G;
batch_size=4096;
enable_profile=off;
```

  |   |   |  
-- | -- | -- | --
  | Base: dev-1.0.1-1c5f7da| After fix |  Time ratio
q1.1 | 347 | 283 | 0.82
q1.2 | 379 | 257 | 0.68
q1.3 | 127 | 132 | 1.04
q2.1 | 3385 | 3148 | 0.93
q2.2 | 1076 | 1056 | 0.98
q2.3 | 863 | 873 | 1.01
q3.1 | 1767 | 1628 | 0.92
q3.2 | 723 | 722 | 1.00
q3.3 | 556 | 541 | 0.97
q3.4 | 108 | 114 | 1.06
q4.1 | 2326 | 2039 | 0.88
q4.2 | 1446 | 1117 | 0.77
q4.3 | 2403 | 1501 | 0.62
Total | 15506 | 13411 | 0.86
  |   | -2095 |  

</byte-sheet-html-origin></span></div>
</div></div>

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
4. Has document been added or modified: (Yes/No/No Need)
5. Does it need to update dependencies: (Yes/No)
6. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yiguolei** [comment]
LGTM
@xinyiZzz Could you please help to review this PR?

**Comment by xinyiZzz** [comment]
> LGTM @xinyiZzz Could you please help to review this PR?

Good job, i will test it on concurrency.

File: be/src/vec/common/allocator.h [hunk 1]
```diff
@@ -146,8 +146,13 @@ class Allocator {
...
                                                           std::to_string(new_size) + ".",
                                                   doris::TStatusCode::VEC_CANNOT_MREMAP);
-
             /// No need for zero-fill, because mmap guarantees it.
+
+            if constexpr (mmap_populate)
+                // MAP_POPULATE seems have no effect for mremap as for mmap,
+                // Zero-fill enlarged memory range explicitly to pre-fault the pages
+                if (new_size > old_size)
+                    memset(reinterpret_cast<char*>(buf) + old_size, 0, new_size - old_size);
```
Comment (xinyiZzz) [inline be/src/vec/common/allocator.h:155]: Why manual zero-fill is needed here,

CK's comment mentions that mmap is automatically zero-filled. The manual also seems to say so: https://linux.die.net/man/2/mmap

Comment (jacktengg) [reply be/src/vec/common/allocator.h:155]: memset explicitly here is for pre-pault the page, not for zero fill. As I tested, MAP_POPULATE seems have no effect for mremap as for mmap.

File: be/src/vec/common/allocator.h [hunk 2]
```diff
@@ -146,8 +146,13 @@ class Allocator {
                                                           std::to_string(old_size) + " to " +
                                                           std::to_string(new_size) + ".",
                                                   doris::TStatusCode::VEC_CANNOT_MREMAP);
-
             /// No need for zero-fill, because mmap guarantees it.
+
+            if constexpr (mmap_populate)
```
Comment (HappenLee) [inline be/src/vec/common/allocator.h:151]: please format the code

**Review by HappenLee — APPROVED** [review]
LGTM

**Comment by xinyiZzz** [comment]
LGTM

File: be/src/vec/common/allocator.h (hunk 1 above)
Comment (xinyiZzz) [reply be/src/vec/common/allocator.h:160]: Thanks for the advice.

**Pull request merged**